agent/src/util/discovery_operator.rs

```diff
@@ -0,0 +1,1411 @@
+                let mut remove_instance = false;
+                match instance_info.connectivity_status {
+                    InstanceConnectivityStatus::Online => {
+                        if is_local {
+                            remove_instance = true;
+                        } else {
+                            let sender = instance_info.list_and_watch_message_sender.clone();
+                            let updated_instance_info = InstanceInfo {
+                                connectivity_status: InstanceConnectivityStatus::Offline(
+                                    Instant::now(),
+                                ),
+                                list_and_watch_message_sender: instance_info
+                                    .list_and_watch_message_sender
+                                    .clone(),
+                            };
+                            self.instance_map
+                                .lock()
+                                .await
+                                .insert(instance.clone(), updated_instance_info);
+                            trace!(
+                                "update_connectivity_status - instance {} went offline ... starting timer and forcing list_and_watch to continue",
+                                instance
+                            );
+                            sender
+                                .send(device_plugin_service::ListAndWatchMessageKind::Continue)
+                                .unwrap();
+                        }
+                    }
+                    InstanceConnectivityStatus::Offline(instant) => {
+                        let time_offline = instant.elapsed().as_secs();
+                        // If instance has been offline for longer than the grace period, terminate the associated device plugin
+                        if time_offline >= SHARED_INSTANCE_OFFLINE_GRACE_PERIOD_SECS {
+                            remove_instance = true;
+                        }
+                    }
+                }
+                if remove_instance {
+                    trace!("update_connectivity_status - instance {} has been offline too long ... terminating device plugin", instance);
+                    device_plugin_service::terminate_device_plugin_service(
+                        &instance,
+                        self.instance_map.clone(),
+                    )
+                    .await
+                    .unwrap();
+                    k8s::try_delete_instance_arc(
+                        kube_interface.clone(),
+                        instance.clone(),
+                        self.config.metadata.namespace.as_ref().unwrap().clone(),
+                    )
+                    .await
+                    .unwrap();
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+pub mod start_discovery {
+    use super::super::registration::DiscoveryHandlerConnectivityStatus;
+    #[double]
```

Comment:
what is `#[double]`?

Reply:
It is a `mockall` that automates importing a mock type in test mode, or a real type otherwise. Added a comment explaining in 901c450 .